---
title: Working with Text using stringr
author: Stephen Pederson
date: '2020-09-28'
slug: working-with-text-using-stringr
categories: []
tags:
  - tidyverse
  - stringr
  - R Markdown
Description: ''
DisableComments: no
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    echo = TRUE,
    results = "hide",
    message = FALSE,
    warning = FALSE,
    fig.align = "center"
    )
```

```{r packages, echo=FALSE}
library(tidyverse)
```

# Introduction

## Key Concepts

Working with text is one of the most common processes in R and we often take it for granted.
Today we'll look at a few important concepts using the package `stringr`.
The key outcomes will be a better understanding of:

- Pattern Matching using *regular expressions*
- Introduction to *categorical variables*, known in R as `factors`
- Learning to build code that is robust to changes in your data

## Setup for Today

To start today, let's return to the same `R Project` as last week.
This should have been called `Improving-R` and if everything was done correctly last week, you'll be able to see this name listed in the drop-down R Project menu in the top-right corner of R Studio.
Just click on the name and you'll be taken to the correct folder and will be ready to go.
**If you missed last week**, please head to the [setup section](https://drmcrl-bioinformatics-training.netlify.app/post/improving-your-data-analysis-in-r/#r-projects) and make sure you have:

1. An R Project with the name `Improving-R`, and 
2. The data file [counts.out](https://raw.githubusercontent.com/UofABioinformaticsHub/Spring_Into_Bioinformatics/master/Day_2/data/2_alignedData/featureCounts/counts.out) saved in the folder `data`.

Once you are in the `Improving-R` R Project, create a new R Markdown file (`File > New File > R Markdown`).
Set the title to be "Working with Text" and add your own name etc.
Save this as `TextManipulation.Rmd`, or some other suitable name.

Find the first chunk (named `setup`) and:

1. Delete everything below the `setup` chunk
2. Add `message = FALSE` inside the function `knitr::opts_chunk$set()`
3. Create a new chunk directly below the first one called `packages`
4. Inside the chunk `packages` add `library(tidyverse)`.

Now compile the document and make sure everything resembles the following screenshot:

![](/images/StringrSetupScreen.png)

# Using `stringr`

The package `stringr` is loaded as part of the `tidyverse` and contains a set of functions which are incredibly useful for manipulating text.
Often we'll need to know these tricks as our sample names may have multiple genotypes or treatments embedded in them, and knowing how to get these pieces of information out can be very helpful.
Other times, we need to know how to correct typos longs the lines of sex being recorded as `male`, `female`, `M`, `m`, `f` or `F` all in the same document.
This type of thing happens more than you may realise, especially when dealing with large datasets that have a lot of manually entered values.
Alternatively we may need to remove a file suffix to make our plots look nicer.

The first thing we'll need to do to get underway, is create a character vector that we'll work with for this session.
Create new chunk called `hi` and create the vector `hi` as follows.

```{r hi}
hi <- c("Hi Mum", "Hi Mother")
```

This is a simple character vector (of length `r length(hi)`) with two closely related phrases.
This will be short enough for us to work with today and see all of the results clearly.
In a real world context, *we often work with vectors with many thousands of entries.* 
These are generally far too long to inspect one element at a time.

The package `stringr` has a whole series of functions, many of which start with `str_`.
From here on, we strongly advise to demonstrate each function within it's own chunk, and for you to make notes in your R Markdown document that help **you** understand what we're doing.
For nearly all of the following examples, we won't actually create a new object, but will just demonstrate the output of each function, which should appear as R output when you compile your document.
Importantly, for the first set of functions we look at, we pass the function a character vector of length `r length(hi)`, and we will return a character vector of length `r length(hi)`.
There are a few exceptions later on though.

## Changing Case

We can very easily change the case of our vector, and this may be of help when our collaborators have given us same strange combination of M/m/F/f encoding for the sex of our samples.
The main functions are `str_to_lower()`, `str_to_upper()` and `str_to_title`, with final helpful function being `str_to_sentence()`.
All of these take one main argument, called `string` in the function.
When we call a function in R, we can name these arguments explicitly, or if we pass everything in order we don't need to name these.
The first two lines of code below are essentially identical.

```{r}
str_to_lower(string = hi)
str_to_lower(hi)
```

Notice that because the first argument expected by `str_to_lower()` is called string, we can name it, or just pass our character vector to the function in the first position.
For the remaining three functions in this set, we will omit the argument name for convenience, although including it is better practice.

```{r}
str_to_upper(hi)
str_to_title(hi)
str_to_sentence(hi)
```

These are quite well named functions as the name makes it clear what the output will be.
This is not always the case with functions, so make sure you describe what each one does.

## Removing Unwanted Text

Often we have to remove unwanted text like file suffixes and directory names, and `str_remove_all()` makes this easy.
There are two arguments we need to provide to this function.
The first is called `string`, whilst the second argument is called `pattern`.
Now, we need to specify two arguments.
We can either name them explicitly, or just pass all of the values in the correct position.

```{r}
str_remove_all(string = hi, pattern = "M")
str_remove_all(hi, "M")
```

### Regular Expression Syntax

In `stringr`, syntax is based on **regular expressions**, so as well as specifying exact patterns to match, we can provide a *set of alternative values*.
In the following, we'll remove either an upper or lower case `m` by providing a set of alternative characters within the square brackets, i.e. `"[Mm]"`.
We'll follow that by removing all lower-case vowels. 

```{r}
str_remove_all(hi, "[Mm]")
str_remove_all(hi, "[aeiou]")
```

We can also set ranges of values to remove, such as all *upper-case* letters, or just those between `a` and `i` in the alphabet.

```{r}
str_remove_all(hi, "[A-Z]")
str_remove_all(hi, "[a-i]")
```

**NB:** There is an alternative function `str_remove()` provided by `stringr` which will only remove the first match to the supplied pattern.
This is mostly not the function you'll be after, but you may choose it by accident at some point, then wonder what went wrong.

#### Wild Cards

Because syntax is based on **regular expressions**, we can also **set a wild-card character by using the `.` symbol.**
For those familiar with regular expressions, this will be old news, but for those who are not familiar with `regexp` syntax this may be a new way of working with text.
Some of you may also be familiar with using the `*` symbol as a plain-text wild-card, but importantly the `*` symbol doesn't work in that manner here and actually *has a completely different meaning*.

```{r}
str_remove_all(hi, "Mu")
str_remove_all(hi, "M.")
```

#### Extending Matches

In this last example, we've removed the upper case `M` and any single character that directly follows it.
We could also remove **any number of following characters** by adding a `+` symbol after the `.`, which is interpreted as *match anything (.) one or more times (+)*

```{r}
str_remove_all(hi, "M.+")
```

Or we can specify the *exact number of matches*.
In this example, we're matching anything (`.`) for exactly `2` characters.

```{r}
str_remove_all(hi, "M.{2}")
```

#### Anchoring Matches

We can also anchor the match to be the beginning or end of a string.
Placing the `^` character at the beginning of a pattern means that the match must occur at the beginning of the string.

```{r}
str_remove_all(hi, "[Hh]")
str_remove_all(hi, "^[Hh]")
```

The dollar symbol (`$`) performs this exact function for matches at the end of a string, and can be placed at the end f the pattern.

```{r}
str_remove_all(hi, "[Mm]")
str_remove_all(hi, "[Mm]$")
```

### Extracting Patterns

As well as just removing sections of a text string,there are various methods for reaching into a text string and grabbing the patterns we need.
The first of these is `str_extract()` which also takes the two arguments `string` and `pattern`.
If we search for a pattern that doesn't exist, `str_extract()` will return a missing value (`NA`).

```{r}
str_extract(hi, "Mu")
str_extract(hi, "M.+")
```

Notice how the patterns behave in the exact same manner as we saw with `str_replace_all()`, by using the `regex` syntax.
We can also use exact positions within a text string using `str_sub()`, with the arguments `string`, `start` and `end`.
Notice that naming the `start` and `end` arguments explicitly is really helpful here as when you read this back in 2 years time, you'll immediately understand what the numbers `4` and `5` represent.

```{r}
str_sub(hi, start = 4, end = 5)
```

## Replacing Text

Sometimes, we don't need to just find or extract one simple pattern, but we need to perform more complex manipulations.
The function `str_replace()` allows us to specify the arguments `string` & `pattern` as we've already seen, but also provide the third argument `replacement`.
Once again, the following two commands are identical, because we are providing the arguments in order.

```{r}
str_replace(string = hi, pattern = "Hi", replacement = "Hello")
str_replace(hi, "Hi", "Hello")
```

We can also use our pattern matching tricks to replace either "Mum" or "Mother" with "Dad".
**Feel free to experiment with other patterns and replacements here.**

```{r}
str_replace(hi, "M.+", "Dad")
```

As well as just using wild-cards, we can specify complete patterns as alternatives.
When searching using this technique, the alternative phrases must be provided within the round braces, and they are separated by the "|" symbol, which is commonly used to represent "OR".

```{r}
str_replace(hi, "(Mum|Mother)", "Dad")
```

### Capturing Text

`str_replace()` also allows us to 'capture' a pattern within each text string, and incorporate it into the replacement.
**To capture a pattern, we surround it in the round braces `()`**, as we did above.
*To return the pattern, we use the shortcut `\\1` which will return the* **first** *pattern we have captured.*
In the following, notice that *we're matching white-space followed by "(M.+)"*, where the captured pattern is "M.+".
In the replacement, we're replacing the entire pattern (including the white-space) with a long string, but this replacement string **includes the captured pattern**.

```{r}
str_replace(
    string = hi, 
    pattern = " (M.+)", 
    replacement = "! We captured and replaced the pattern \\1."
)
```

We can also capture multiple patterns, and return them in any order we choose.
In the following:

- our first capture will be "Hi"
- then we *don't capture the white-space*, but
- then we capture the pattern "M.+" as we did before.

In our replacement, we're returning them in the opposite order that we captured them in, and we're also including some exclamation marks and additional white-space.

```{r}
str_replace(hi, "(Hi) (M.+)", "\\2! \\1!")
```

Just like `str_remove()` and `str_remove_all()`, there are two versions of `str_replace()`.
The first will only replace the first match to the pattern, whilst `str_replace_all()` will replace all matches to the pattern.
Depending on what operation you're needing to perform, either may be suitable. 

```{r}
str_replace(hi, "[Mm]", "b")
str_replace_all(hi, "[Mm]", "b")
```

## A few more handy functions

The above functions are probably the "big guns" of `stringr`, however there are a few more very useful things we can do to clean up our data.
If we wish to ensure that our text strings *are all the same length*, we can pad them with any character we choose, using `str_pad()`.
This takes the arguments `string`, `width`, `side` and `pad` as the four arguments.
In the first example, we're setting our new strings to be exactly 10 characters wide, and we're padding on the right with exclamation marks.
Note again, that explicitly naming the arguments here helps us understand exactly what we're doing with the function.

```{r}
str_pad(hi, width = 10, side = "right", pad = "!")
```

Another common use for something like this would be to pad numbers to ensure we have the same number of digits.
If numbers are represented as characters, most sorting algorithms sort them alpha-numerically instead of numerically.
We often see strange ordering like `r pander::pander(sort(as.character(1:10)))` instead of the more reasonable `r pander::pander(1:10)`.
In the next line, we'll add zeroes in front of the numbers to ensure every number has two digits.
Once we've done this, alpha-numeric sorting and numeric sorting will give the same results.

```{r}
str_pad(1:10, width = 2, pad = "0")
```

Notice that we didn't set the argument `side` here!
By default, this argument will be set to `side = "left"`, and we've relied on this value.
This means that `pad = "0"` will be provided in position 3 of the function, so **must** be explicitly named.
This also demonstrates why it can be helpful to explicitly name arguments.

Sometimes we can end up with ridiculously long text strings, particularly when dealing with gene-set enrichment analyses, so we can use `str_trunc()` to truncate these. 
By default, the last 3 characters of a truncated string will be given as the ellipsis (`...`) to indicate the text has been shortened.

```{r}
str_trunc(hi, width = 6)
```

Finally, we often see text strings which have excessive white-space, so we can remove this using `str_trim()`.
This will remove leading or trailing white-spaces from the text.

```{r}
str_trim("   Hi!")
```

## Approaches That Return Different Structures

In all of the above, we provided a character vector of length `r length(hi)` and we obtained a character vector of length `r length(hi)` as our output.
(`NA` technically still counts as a character.)
There are multiple other functions which exist that return output in different forms

### Finding Matching Text

The first of these might be to perform a logical test for the presence of a pattern within our text strings.
`str_detect()` again takes the two arguments `string` and `pattern`, with a third argument called `negate` that defaults to `FALSE`.
If we set this to `TRUE`, we will invert our results, which is often referred to as negating a logical search.
In both cases, we no longer return a character vector, but return the logical values `TRUE` and `FALSE`.

```{r}
str_detect(hi, "Mum")
str_detect(hi, "Mum", negate = TRUE)
```

These functions really come into their own when we're dealing with spreadsheet-like structures (known as data frames) and we wish to perform some kind of filtering on our values, as we would with Auto-Filter in Excel.

Sometimes it can be convenient to look in specific positions within our text and the two functions `str_starts()` and `str_ends()` check to see if a string starts or ends with a specific pattern.
Both of these also take the argument `negate`, which we'll leave as the default (`negate = FALSE`).

```{r}
str_starts(hi, "Hi")
str_ends(hi, "m")
```

We can also subset our initial character vector, so that only the elements which match our search pattern are returned.
Here, we provide a character vector of length `r length(hi)` and initially return a character vector of length `r length(str_subset(hi, "Mum"))`.
By incorporating `regex` wildcards, we can then return any elements that contains "M.".

```{r}
str_subset(hi, "Mum")
str_subset(hi, "M.")
```

A final function (`str_view_all()`) returns a completely different output.
This matches the specified pattern and opens the results in the **Viewer** pane, which we haven't used until now.
This can be incredibly helpful if you're trying to build a complex search pattern and want to check where it's matching. 

```{r, eval=FALSE}
str_view_all(hi, "M.+")
```

### Counting letters

Sometimes, it can be helpful to obtain information about the content of our text strings, such as their overall length, or the number of times a specific letter or pattern appear.
I'm sure you'll immediately realise this can be very useful if wanting to assess the nucleotide content of a DNA/RNA sequence.
In NextGen Sequencing experiments, the GC content of a sequence is a known factor that biases our ability to sequence a DNA fragment.
In all of these, we'll provide our input as a character vector with length `r length(hi)`, but will obtain an integer vector of the same length as the output.

```{r}
str_length(hi)
str_count(hi, "[Hh]")
str_count("AGCTGCGCGATTTAGC", "[GC]")
```

### Collapsing and Joining Text

Sometimes, we need to condense multiple text strings into a single string, and we can use `str_flatten` for this.

```{r}
str_flatten(hi, collapse = ", ")
```

However, there are two `base` functions `paste()` and `paste0()` that are just as helpful here, and probably a little more flexible.

```{r}
paste(hi, collapse = ", ")
```

In addition to performing the above concatenation, we more commonly use the argument `sep` within the function `paste()` to join additional text to a string.
In the following, we're pasting an additional phrase to each of our initial vector elements, and specifying the separator between the phrases to be `", "`.

```{r}
paste(hi, "I hope you're well.", sep = ", ")
```

The function `paste0()` is a simple wrapper to `paste()` but with `sep = ""` as the default.

```{r}
paste0(hi, "!")
```

## Separating Text

Finally, a common operation is to separate a single text string into multiple text strings based on a specific character, or set of characters.
The function that gives us the most convenient output for this is `str_split_fixed()`, where we provide the arguments `string`, `pattern` and `n`, with the final argument controlling how many strings to obtain from the initial string.
In the following, we're splitting our text based on white-space, and ensuring that each string is only split into 2 new strings.
This provides output in a structure known as a `matrix`, which we'll explore further in the course.

```{r}
str_split_fixed(hi, " ", 2)
```

In general, when performing this type of operation, we're looking for something specific and we'll know what type of output we need.
There is a slightly more flexible function `str_split()`, however the output format from `str_split()` is a `list` so a working knowledge of list objects in R is really essential for using this function.

```{r}
str_split(hi, " ")
```

# Using This In Context

## Loading The Counts

The above was a fairly detailed introduction to manipulating text, but to really understand things we should perform these operations in a contact that we're familiar with.
Let's load the data object that we prepared last week and use our new tricks on the sample names.

If you can recall, we loaded this object using the following code.

```{r, echo=FALSE}
counts <- "https://raw.githubusercontent.com/UofABioinformaticsHub/Spring_Into_Bioinformatics/master/Day_2/data/2_alignedData/featureCounts/counts.out" %>%
    read_tsv(comment = "#")
```

```{r, eval=FALSE}
fl <- here::here("data/counts.out")
counts <- read_tsv(fl, comment = "#")
```

As we discussed last week, this is the output from `featureCounts` and we have the following columns.

1. `Geneid`: This is clearly the gene identifier, here we have Ensembl
2. `Chr`: The chromosome, with one value for each exon
3. `Start`: The start positions for each exon
4. `End`: The end positions for each exon
5. `Strand`: The strand of each exon
6. `Length`: The combined length of all exons combined
7. The first column with actual counts from an RNA-Seq experiment

If setting this up for an RNA-Seq analysis using `edgeR` we would need to form a numeric matrix with

- Gene IDs as row names
- Sample IDs as column names

As this manipulation is not the focus of the current session, my strategy for doing this might be as follows.

```{r}
counts <- counts %>%
    dplyr::select(Geneid, ends_with("bam")) %>%
    rename_all(basename) %>%
    rename_all(str_remove_all, pattern = ".trimmed.+") %>%
    column_to_rownames("Geneid") %>%
    as.matrix()
```

The steps taken here were all chained together using the `magrittr` and consisted of

1. Select only the relevant columns. Notice I called by name not position. This is very important as **if you run `featureCounts` on the same set of bam files twice, the column order will not be the same in both files**
2. Removed any directory prefix from all column names using the function `basename()`
3. Use `str_replace_all()` to remove `.trimmedAligned.sortedByCoord.out.bam` from the end of each sample 
4. Shifted the `Geneid` column to be rownames
5. Converted into a `matrix`

Alternatively, if we wish to edit as we parse the data into R.
The following approach is an alternative that ensures we never have the same object in two forms within our workspace.

```{r, eval=FALSE}
counts <- read_tsv(fl, comment = "#") %>%
    dplyr::select(Geneid, ends_with("bam")) %>%
    rename_all(basename) %>%    
    rename_all(str_remove_all, pattern = ".trimmed.+") %>%
    column_to_rownames("Geneid") %>%
    as.matrix()  
```

## Creating a Metadata Object

Sample names often contain the key information we need and in R, we can obtain this information fairly easily using our new skills.
First we can start by forming a `tibble` with a single column.
This gives us a chance to see where we are before we go any further.

```{r}
tibble(
    file = colnames(counts)
)
```

There are two pieces of information here which are important:

1. The SRA Identifier
2. The tissue, annotated as either `cbc` or `skm` for cerebral cortex or skeletal muscle respectively.

Let's try to get these into separate columns, starting with the SRA ID

```{r}
tibble(
    file = colnames(counts)
) %>%
    mutate(
        sra_id = str_extract(file, "SRR[0-9]+")
    )
```

Now let's extract the tissue

```{r}
tibble(
    file = colnames(counts)
) %>%
    mutate(
        sra_id = str_extract(file, "SRR[0-9]+"),
        tissue = str_extract(file, "(cbc|skm)")
    )
```

An important perspective here is that the SRA id is a text string which identifies a sample.
In contrast, the tissue is a categorical variable that we will include in our statistical model.
By default, `R` will convert any character columns to factors during analysis, and will always set these in alphabetic order.
Here, the default would be `cbc` as the baseline (or reference) with `skm` as the second level.
A general principle is that we wish to specify our baseline (or control) group as the first (i.e. reference) factor level.
This isn't really important for this dataset, but if you had DHT and Veh, you'd want to set Veh as the reference level.
Relying on R to do this in alphabetic order may not always be wise.

Let's manually specify the factor levels.

```{r}
tibble(
    file = colnames(counts)
) %>%
    mutate(
        sra_id = str_extract(file, "SRR[0-9]+"),
        tissue = str_extract(file, "(cbc|skm)"),
        tissue = factor(tissue, levels = c("cbc", "skm"))
    )
```

Notice that the column now contains a factor whilst the first two columns are printed as characters.
This will not only impact the interpretation of our statistical model, but can also impact the order values are shown in when plotting.

The `tidyverse` package `forcats` has a number of useful function for working with factors, with a possible alternative here being `fct_inorder()`.
I personally like to be explicit though.
If a file was somehow included without `cbc` or `skm` an `NA` would result and this would (hopefully) alert me that the data had taken an unexpected form.

Maybe we can add a better description of the tissue for our collaborators to understand the data.

```{r}
tibble(
    file = colnames(counts)
) %>%
    mutate(
        sra_id = str_extract(file, "SRR[0-9]+"),
        tissue = str_extract(file, "(cbc|skm)"),
        tissue = factor(tissue, levels = c("cbc", "skm")),
        description = case_when(
            str_detect(file, "cbc") ~ "Cerebral Cortex",
            str_detect(file, "skm") ~ "Skeletal Muscle"
        )
    )
```

This looks pretty good to me, so at this point I would save this as a metadata object.

```{r}
samples <- tibble(
    file = colnames(counts)
) %>%
    mutate(
        sra_id = str_extract(file, "SRR[0-9]+"),
        tissue = str_extract(file, "(cbc|skm)"),
        tissue = factor(tissue, levels = c("cbc", "skm")),
        description = case_when(
            str_detect(file, "cbc") ~ "Cerebral Cortex",
            str_detect(file, "skm") ~ "Skeletal Muscle"
        )
    )
```

If we're adventurous we might also like to add the library sizes, and although this isn't text manipulation, it's good to see here.

```{r}
samples <- tibble(
    file = colnames(counts)
) %>%
    mutate(
        sra_id = str_extract(file, "SRR[0-9]+"),
        tissue = str_extract(file, "(cbc|skm)"),
        tissue = factor(tissue, levels = c("cbc", "skm")),
        description = case_when(
            str_detect(file, "cbc") ~ "Cerebral Cortex",
            str_detect(file, "skm") ~ "Skeletal Muscle"
        ),
        library_size = colSums(counts)[file]
    )
```

Now that we have our metadata, I'll often summarise this into a table.
The package `pander` is extremely useful for this, so please add `library(pander)` to your `packages` chunk *and make sure you execute the chunk*.

```{r}
library(tidyverse)
library(pander)
```


Then add the following.

```{r}
samples %>%
    group_by(description) %>%
    summarise(
        n = n(),
        `Ave. Library Size` = mean(library_size)
    ) %>%
    rename_all(str_to_title) %>%
    pander(
        caption = "Summary of RNA-Seq Counts",
        big.mark = ","
    )
```

Notice that we've used `str_to_title()` inside the call to `rename_all()` to make our table look nice, but without changing our data object.


# Conclusion

In the above you may have noticed that nothing was ever called by position using `x[1:5]` or something similar.
This helps build good practice.
If we call by position (e.g. `x[1:5]`) we create two problems:

1. If the data structure changes our analysis may break, **or worse**, may complete without us being aware of the changes producing invalid results
2. When we revisit our code in the future, we won't know what's in any object when selecting by position. If we select columns 1:5, we have to remember what's in those columns and check that everything is correct.

When forming the counts object, I used `dplyr::select()` and when reading that back, it's very clear what we did.

```{r eval=FALSE}
dplyr::select(counts, Geneid, ends_with("bam"))
```

is much easier to interpret than

```{r, eval = FALSE}
counts[,c(1, 7:14)]
```

When we defined the metadata, we pulled all values directly from the `counts` object, ensuring that if the order changed our code will still produce the object in the correct structure.

The *old-school* method of defining the tissue might have been

```{r}
rep(c("cbc", "skm"), times = 4)
```

This is not robust to the order changing, or any samples being added/removed.
Using `str_extract()` is incredibly simple, but makes for code which is far more robust to changes in your data structure.
As mentioned, `featureCounts` will produce columns in a different order every time it runs on the same set of bam files.
Building code in the way we have done, protects us from this kind of problem.




